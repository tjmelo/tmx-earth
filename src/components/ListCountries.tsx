import React, { useEffect, useState } from 'react'
import { apiListCountries } from '../model/api'
import { useDispatch } from 'react-redux'
import { update } from '../feature/country/countrySlice'
import Loading from './Load'

import { ICommonName, ITargetEvent } from '../interfaces'

export const ListCountries = () => {
  const dispatch = useDispatch()
  const [data, setData] = useState<[]>([])

  useEffect(() => {
    apiListCountries().then(({ data, status }) => 
      status !== 200 ? setData([]) : setData(data))
  }, [])

  const selectCountry = (event: ITargetEvent) => {
    dispatch(update(event.target.value))
  }

  const OptionsCountry = ({name}: ICommonName, idx: number) => (
    <option key={idx}>{name.common}</option>
  )

  return (
    <div className="container">
      {!data 
        ? <Loading type='danger'>Data not found, notify the administrator!</Loading> 
        : <select
            className="form-select form-select-lg my-2 shadow"
            onChange={selectCountry}
          >
            <option value="Select a country">Select a country</option>
            {data && data.map(OptionsCountry)}
          </select>
      }
    </div>
  )
}
