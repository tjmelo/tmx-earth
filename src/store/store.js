// Require configureStore from redux toolkit...
import { configureStore } from '@reduxjs/toolkit';

// Require createSlice from countryReducer...
import countryReducer from '../feature/county/countrySlice';

// Export method configureStore with object reducer...
export default configureStore({
    reducer: {
        country: countryReducer,
    }
})