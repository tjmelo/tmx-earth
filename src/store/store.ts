// Require configureStore from redux toolkit...
import { configureStore } from '@reduxjs/toolkit';

// Require createSlice from countryReducer...
import countryReducer from '../feature/country/countrySlice';

// Export method configureStore with object reducer...
export default configureStore({
    reducer: {
        country: countryReducer,
    }
})