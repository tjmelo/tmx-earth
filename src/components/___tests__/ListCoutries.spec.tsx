import {fireEvent, render, screen, waitFor} from '@testing-library/react'
import '@testing-library/jest-dom'

import { ListCountries } from '../ListCountries'
import { Provider } from 'react-redux'
import store from '../../store/store'

test('Should render List coutries options', () => {
    //do
    const {asFragment, getByText} = render(
        <Provider store={store}>
            <ListCountries />
        </Provider>
    )

    // then
    expect(getByText('Select a country')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot();
})

test('Should render another option to select', () => {
    //do
    render(
        <Provider store={store}>
            <ListCountries />
        </Provider>
    )
    
    const optionSelectTest = 'Brazil'
    const input: HTMLSelectElement = screen.getByLabelText('select-country')

    fireEvent.change(input.options[0], {target: {value: optionSelectTest}})
    expect(input.value).toBe(optionSelectTest)

})
