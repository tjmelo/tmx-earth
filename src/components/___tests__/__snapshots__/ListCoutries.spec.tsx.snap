// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render List coutries options 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <select
      aria-label="select-country"
      class="form-select form-select-lg my-2 shadow"
    >
      <option
        value="Select a country"
      >
        Select a country
      </option>
    </select>
  </div>
</DocumentFragment>
`;
